### **Nombre:** Elda Magally Calderón Motta
### **Carné:** 16003182
### **Sección:** U

# **TAREA 3**

# Maldición de Dimensionalidad

In [214]:
import numpy as np
from sklearn.preprocessing import PolynomialFeatures
from itertools import combinations

## Una variable

In [215]:
X = np.arange(4).reshape(-1,1)
X

array([[0],
       [1],
       [2],
       [3]])

In [216]:
poly = PolynomialFeatures(degree=3)

In [217]:
poly.fit_transform(X)

array([[ 1.,  0.,  0.,  0.],
       [ 1.,  1.,  1.,  1.],
       [ 1.,  2.,  4.,  8.],
       [ 1.,  3.,  9., 27.]])

In [218]:
poly.get_feature_names(['a'])

['1', 'a', 'a^2', 'a^3']

In [219]:
a = len(poly.get_feature_names(['a']))

In [220]:
b = poly.get_feature_names(['a', 'b'])

In [221]:
i = 0
for j in range(1, a+1):
    for c in combinations(b, j):
        if ('a^3') in c:
            i = i+1
            print(i, c)

1 ('a^3',)
2 ('1', 'a^3')
3 ('a', 'a^3')
4 ('a^2', 'a^3')
5 ('1', 'a', 'a^3')
6 ('1', 'a^2', 'a^3')
7 ('a', 'a^2', 'a^3')
8 ('1', 'a', 'a^2', 'a^3')


Con una variable se podrían generar 8 polinomios diferentes de grado 3

## Dos variable

In [222]:
X = np.arange(4).reshape(2,2)
X

array([[0, 1],
       [2, 3]])

In [223]:
poly = PolynomialFeatures(degree=3)

In [224]:
poly.fit_transform(X)

array([[ 1.,  0.,  1.,  0.,  0.,  1.,  0.,  0.,  0.,  1.],
       [ 1.,  2.,  3.,  4.,  6.,  9.,  8., 12., 18., 27.]])

In [225]:
poly.get_feature_names(['a', 'b'])

['1', 'a', 'b', 'a^2', 'a b', 'b^2', 'a^3', 'a^2 b', 'a b^2', 'b^3']

In [226]:
a = len(poly.get_feature_names(['a', 'b']))

In [227]:
b = poly.get_feature_names(['a', 'b'])

In [228]:
i = 0
for j in range(1, a+1):
    for c in combinations(b, j):
        if ('a^3' or 'b^3' or 'a^2 b' or 'a b^2') in c:
            i = i+1
print(i, c)

512 ('1', 'a', 'b', 'a^2', 'a b', 'b^2', 'a^3', 'a^2 b', 'a b^2', 'b^3')


Con dos variables se podrían generar 512 polinomios diferentes de grado 3

## Tres variable

In [229]:
X = np.arange(6).reshape(2,3)
X

array([[0, 1, 2],
       [3, 4, 5]])

In [230]:
poly = PolynomialFeatures(degree=3)

In [231]:
poly.fit_transform(X)

array([[  1.,   0.,   1.,   2.,   0.,   0.,   0.,   1.,   2.,   4.,   0.,
          0.,   0.,   0.,   0.,   0.,   1.,   2.,   4.,   8.],
       [  1.,   3.,   4.,   5.,   9.,  12.,  15.,  16.,  20.,  25.,  27.,
         36.,  45.,  48.,  60.,  75.,  64.,  80., 100., 125.]])

In [232]:
poly.get_feature_names(['a', 'b', 'c'])

['1',
 'a',
 'b',
 'c',
 'a^2',
 'a b',
 'a c',
 'b^2',
 'b c',
 'c^2',
 'a^3',
 'a^2 b',
 'a^2 c',
 'a b^2',
 'a b c',
 'a c^2',
 'b^3',
 'b^2 c',
 'b c^2',
 'c^3']

In [233]:
a = len(poly.get_feature_names(['a', 'b', 'c']))

In [234]:
b = poly.get_feature_names(['a', 'b', 'c'])

In [237]:
i = 0
for j in range(1, a+1):
    for c in combinations(b, j):
        if ('a^3' or 'b^3' or 'c^3' or 'a^2 b' or 'a b^2' or 'a^2 c' or 'a b c' or 'a c^2' or 'b^2 c' or 'b c^2') in c:
            i = i+1
print(i, c)

524288 ('1', 'a', 'b', 'c', 'a^2', 'a b', 'a c', 'b^2', 'b c', 'c^2', 'a^3', 'a^2 b', 'a^2 c', 'a b^2', 'a b c', 'a c^2', 'b^3', 'b^2 c', 'b c^2', 'c^3')


Con tres variables se podrían generar 524 288 polinomios diferentes de grado 3

# Conclusiones

Para experimentar un poco con la maldición de dimensionalidad se produjeron los diferentes polinomios de grado 3 que se pueden generar al tener 1, 2 o 3 variables.

Al ejecutar los experimentos se puede observar como al aumentar la dimensionalidad de 1 hasta 3 variables las combinaciones de estas tienen un aumento sumamente significativo, pasando desde 8 a 512 y finalmente 524288 polinomios diferentes con las combinaciones de las variables